### ISSUE TITLE
Will  DADI support device mapper in later?

### ISSUE THREAD

**Issue by wxx213** [containerd/accelerated-container-image#109]
The the rootfs size limit will prevent one of the containers in the same host cost too much disk space.

It seems that the overlayfs cannot make it. The device mapper may help, but I don't know if  DADI could support device mapper.

**Comment by BigVan** [comment]
We have no plan to support device-mapper because DM management is complex and create/delete a thin volume on snapshot preparing is much slower than overlayfs [see it](https://github.com/containerd/containerd/discussions/6625).

Overlaybd is a stand alone storage driver which can provide Rootfs without using overlayfs directly. 

There are 2 ways to realize disk quota:
1. Choose the properly disk size you want when making image's base layer
2. Use command like 'tune2fs -O project,quota' to limit /dev/sdX.

**Comment by liulanzheng** [comment]
I think it is about the disk quota of writable layer? @BigVan Is there any method to set quota for the upper dir?

**Comment by BigVan** [comment]
> I think it is about the disk quota of writable layer? @BigVan Is there any method to set quota for the upper dir?

I remember it can be limited by tune2fs and I have successfully done it before. Let me try it later.

**Comment by wxx213** [comment]
> We have no plan to support device-mapper because DM management is complex and create/delete a thin volume on snapshot preparing is much slower than overlayfs [see it](https://github.com/containerd/containerd/discussions/6625).
> 
> Overlaybd is a stand alone storage driver which can provide Rootfs without using overlayfs directly.
> 
> There are 2 ways to realize disk quota:
> 
> 1. Choose the properly disk size you want when making image's base layer
> 2. Use command like 'tune2fs -O project,quota' to limit /dev/sdX.

What do you mean for the disk to make image's base layer in the first one, should I make some config when convert the image?

**Comment by wxx213** [comment]
> We have no plan to support device-mapper because DM management is complex and create/delete a thin volume on snapshot preparing is much slower than overlayfs [see it](https://github.com/containerd/containerd/discussions/6625).
> 
> Overlaybd is a stand alone storage driver which can provide Rootfs without using overlayfs directly.
> 
> There are 2 ways to realize disk quota:
> 
> 1. Choose the properly disk size you want when making image's base layer
> 2. Use command like 'tune2fs -O project,quota' to limit /dev/sdX.

@BigVan What do you mean for the disk to make image's base layer in the first one, should I make some config when convert the image?

**Comment by liulanzheng** [comment]
@wxx213 If you want to change the disk virtual size, for now you have to change the code. I think changing the virtual block size of the image disk makes no sense, because it is read only and cost the disk space of cache. So I need to find out exactly which part costs too much disk space.

**Comment by wxx213** [comment]
> @wxx213 If you want to change the disk virtual size, for now you have to change the code. I think changing the virtual block size of the image disk makes no sense, because it is read only and cost the disk space of cache. So I need to find out exactly which part costs too much disk space.

Yes, I also made a test, the container rootfs size only depends the mountpoint of the overlayfs for container root.
And I'm trying the quota feature of xfs and ext4. May be DADI could add support for the quota to isolate container's rootfs which is really meaningful in production env.
